### PR TITLE
[SPARK-15140][SPARK-15441][SQL][WIP] support null object in encoder

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently we can't encoder top level null object into internal row, it throws NPE in 1.6 and returns incorrect result in 2.0.

The root cause is: Spark SQL doesn't allow row to be null, only its columns can be null. This is different from objects, object itself can be null, and its fields can also be null.

This is not a problem before, as we assume the input object is never null. However, for outer join, we do need the semantic of null object.

This PR tries to resolve this fundamental problem by adding a hidden column when serialize object to row, to indicate if the object is null or not.
## How was this patch tested?

existing test and new test in `DatasetSuite`

TODO: add more comments, code cleanup

### CODE REVIEW

**Comment by davies** [comment]
@cloud-fan Maybe it's not that easy to propogate the special column all the way down, we could just use this trick to fix the outer join issue?

**Comment by cloud-fan** [comment]
@davies , to add this trick to outer-join, we still need to improve our encoder framework to support it. So in this PR I added the infrastructure to encoder framework but only use it for outer join.

**Comment by cloud-fan** [comment]
cc @yhuai @marmbrus

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 1]
```diff
@@ -763,6 +763,22 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+    assert(result(0) == 1 -> "a")
+    assert(result(1) == null)
+  }
+
+  test("SPARK-15441: Dataset outer join") {
+    val left = Seq(ClassData("a", 1), ClassData("b", 2)).toDS().as("left")
+    val right = Seq(ClassData("x", 2), ClassData("y", 3)).toDS().as("right")
+    val joined = left.joinWith(right, $"left.b" === $"right.b", "left")
+    joined.explain(true)
+    joined.show()
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: Should we check the result?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 2]
```diff
@@ -170,6 +174,17 @@ object ExpressionEncoder {
       e4: ExpressionEncoder[T4],
       e5: ExpressionEncoder[T5]): ExpressionEncoder[(T1, T2, T3, T4, T5)] =
     tuple(Seq(e1, e2, e3, e4, e5)).asInstanceOf[ExpressionEncoder[(T1, T2, T3, T4, T5)]]
+
+  private val nullFlagName = "is_null_obj"
+  private val nullFlagMeta = new MetadataBuilder().putNull(nullFlagName).build()
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: putBoolean?

**Comment by zhzhan** [comment]
My understanding is that this new added hidden column is mainly for serdes object to/from row. How would you leverage it to solve the the out join case where the null object is actually added during query execution?

**Comment by cloud-fan** [comment]
@zhzhan , for outer join, if it's a typed join, then both of the join side will include this hidden column, as their columns are produced by serializing custom object. If we need to return null for one join side, we will null out all columns of that side, including the hidden column. Then when we deserialize it, we can know that this join side is null and we can build a null object for it.

**Comment by cloud-fan** [comment]
closing in favor of https://github.com/apache/spark/pull/13425

**Pull request closed without merging**